File: packages/python-host/src/ai37_agent_host/metrics.py
"""Prometheus-метрики хоста — зеркало ``ts-host/src/metrics.ts`` (те же имена/лейблы, чтобы
Tier-2 алерты были едины для TS- и Python-агентов).

Единый default-REGISTRY (``make_asgi_app()`` в ``create_agent_host`` отдаёт именно его),
ТОЛЬКО низкокардинальные ``ai37_*``-серии (лейблы ограничены service/transport/status/
final_state/reason — никаких per-user / per-context / per-request значений). Скрейпится
внутрикластерным Alloy'ем по ``GET /metrics`` (includeMetrics=``ai37_.*``).

Все инкременты обёрнуты в ``_safe`` и НИКОГДА не бросают — сбой метрик не должен ломать ход.
"""

from __future__ import annotations

import re
from collections.abc import Callable

from prometheus_client import Counter, Histogram

_requests_total = Counter(
    "ai37_agent_requests_total",
    "Total agent turns handled by the host, by transport and outcome (ok|error).",
    ["service", "transport", "status"],
)
_request_duration = Histogram(
    "ai37_agent_request_duration_seconds",
    "Agent turn duration in seconds, by transport.",
    ["service", "transport"],
    buckets=(0.1, 0.5, 1, 2, 5, 10, 30, 60, 120),
)
_tasks_total = Counter(
    "ai37_agent_tasks_total",
    "Terminal task states produced by the host (completed|failed|input_required).",
    ["service", "final_state"],
)
_billing_denied_total = Counter(
    "ai37_billing_denied_total",
    "Billing preflight denials surfaced by the host, by reason.",
    ["service", "reason"],
)
_auth_failures_total = Counter(
    "ai37_agent_auth_failures_total",
    "401 responses from the host auth guard.",
    ["service"],
)


def norm_final_state(status: str) -> str:
    """AgentResult.status → label-safe (``input-required`` → ``input_required``)."""
    return "input_required" if status == "input-required" else status


def service_label(name: str | None) -> str:
    """``service``-лейбл из build_info['service']: строчный slug ``[a-z0-9_-]`` ≤63 симв.

    Значение ФИКСИРОВАНО на процесс (одно на деплой) → кардинальность лейбла = 1.
    """
    slug = re.sub(r"[^a-z0-9_-]+", "-", (name or "unknown").lower()).strip("-")[:63]
    return slug or "unknown"


def _safe(fn: Callable[[], None]) -> None:
    try:
        fn()
    except Exception:  # noqa: BLE001 - метрики не должны ломать ход агента
        pass


def observe_turn(service: str, transport: str, final_state: str, seconds: float) -> None:
    """Один ход (turn) на транспорте ``a2a``/``agui``: rate + errors + duration + terminal-state."""

    def _do() -> None:
        status = "error" if final_state == "failed" else "ok"
        _requests_total.labels(service, transport, status).inc()
        _request_duration.labels(service, transport).observe(seconds)
        _tasks_total.labels(service, final_state).inc()

    _safe(_do)


def record_billing_denied(service: str, reason: str) -> None:
    """Отказ биллинг-preflight (``BillingExecutionDeniedError.reason``)."""
    _safe(lambda: _billing_denied_total.labels(service, reason).inc())


def record_auth_failure(service: str) -> None:
    """401 из AuthGuardMiddleware (сбой Authentik/JWKS/верификатора)."""
    _safe(lambda: _auth_failures_total.labels(service).inc())

File: packages/ts-host/src/metrics.ts
import { Registry, Counter, Histogram } from 'prom-client'
import type { AgentStatus } from './types'

/**
 * Prometheus-метрики хоста. Единый Registry, ТОЛЬКО низкокардинальные `ai37_*`-серии
 * (лейблы ограничены service/transport/status/final_state/reason — никаких per-user /
 * per-context / per-request значений). Скрейпится внутрикластерным Alloy'ем по `GET /metrics`
 * (k8s-monitoring annotationAutodiscovery, includeMetrics=`ai37_.*`).
 *
 * `collectDefaultMetrics()` НЕ вызываем: process_/nodejs_ не проходят allowlist и лишь раздули
 * бы бюджет активных серий. Все инкременты обёрнуты в `safe()` и НИКОГДА не бросают — сбой
 * метрик не должен ломать ход агента.
 */
export const registry = new Registry()

const requestsTotal = new Counter({
  name: 'ai37_agent_requests_total',
  help: 'Total agent turns handled by the host, by transport and outcome (ok|error).',
  labelNames: ['service', 'transport', 'status'] as const,
  registers: [registry],
})

const requestDuration = new Histogram({
  name: 'ai37_agent_request_duration_seconds',
  help: 'Agent turn duration in seconds, by transport.',
  labelNames: ['service', 'transport'] as const,
  buckets: [0.1, 0.5, 1, 2, 5, 10, 30, 60, 120],
  registers: [registry],
})

const tasksTotal = new Counter({
  name: 'ai37_agent_tasks_total',
  help: 'Terminal task states produced by the host (completed|failed|input_required).',
  labelNames: ['service', 'final_state'] as const,
  registers: [registry],
})

const billingDeniedTotal = new Counter({
  name: 'ai37_billing_denied_total',
  help: 'Billing preflight denials surfaced by the host, by reason.',
  labelNames: ['service', 'reason'] as const,
  registers: [registry],
})

const authFailuresTotal = new Counter({
  name: 'ai37_agent_auth_failures_total',
  help: '401 responses from the host JWT guard.',
  labelNames: ['service'] as const,
  registers: [registry],
})

/** Нормализует AgentStatus в label-safe значение (`input-required` → `input_required`). */
export function normFinalState(status: AgentStatus): string {
  return status === 'input-required' ? 'input_required' : status
}

/**
 * `service`-лейбл из имени агент-карты: строчный slug из `[a-z0-9_-]`, обрезанный до 63 симв.
 * Значение ФИКСИРОВАНО на процесс (одно на деплой) → кардинальность лейбла = 1.
 */
export function serviceLabel(name: string | undefined): string {
  const s = (name ?? 'unknown')
    .toLowerCase()
    .replace(/[^a-z0-9_-]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 63)
  return s || 'unknown'
}

function safe(fn: () => void): void {
  try {
    fn()
  } catch {
    /* метрики не должны ломать ход агента */
  }
}

/** Один ход (turn) на транспорте `a2a`/`agui`/`mcp`: rate + errors + duration + terminal-state. */
export function observeTurn(
  service: string,
  transport: 'a2a' | 'agui' | 'mcp',
  finalState: string,
  seconds: number,
): void {
  safe(() => {
    requestsTotal.inc({ service, transport, status: finalState === 'failed' ? 'error' : 'ok' })
    requestDuration.observe({ service, transport }, seconds)
    tasksTotal.inc({ service, final_state: finalState })
  })
}

/** Отказ биллинг-preflight (BillingExecutionDeniedError.reason). */
export function recordBillingDenied(service: string, reason: string): void {
  safe(() => billingDeniedTotal.inc({ service, reason }))
}

/** 401 из jwtGuard (сбой Authentik/JWKS/верификатора). */
export function recordAuthFailure(service: string): void {
  safe(() => authFailuresTotal.inc({ service }))
}

/** Content-Type для ответа `GET /metrics` (Prometheus text exposition). */
export const metricsContentType = registry.contentType

/** Рендер экспозиции метрик для `GET /metrics`. */
export async function renderMetrics(): Promise<string> {
  return registry.metrics()
}
